src/Data/Parameterized/MonoidSolver.hs

```diff
@@ -0,0 +1,210 @@
+{-|
+Description : A solver for type-level equations in monoids
+Copyright   : (c) Galois, Inc 2021
+Maintainer  : Langston Barrett
+
+Implementation of section 6 of \"A well-known representation of monoids and its
+application to the function \'vector reverse\'\" by Wouter Swierstra.
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Data.Parameterized.MonoidSolver
+  ( solve
+  , solveNat
+  ) where
+
+import Data.Kind (Type)
+import Data.Proxy (Proxy(..))
+import Data.Type.Equality (type (:~:)(Refl))
+import GHC.TypeLits (Nat)
+
+import Data.Parameterized.NatRepr (NatRepr, type (+), plusAssoc)
+
+-- | 'Op' corresponds to @Add@ from the paper.
+type family Op k (m1 :: k) (m2 :: k) :: k
+-- | 'Unit' corresponds to @Add@ from the paper.
+type family Unit k :: k
+
+type instance Op Nat n1 n2 = n1 + n2
+type instance Unit Nat = 0
+
+-- | Type level only
+data MonoidExpr k
+  = EUnit
+  | EVar k
+  | EOp (MonoidExpr k) (MonoidExpr k)
+
+data MERepr (k :: Type) (f :: k -> Type) (me :: MonoidExpr k) where
+  EUnitRepr :: MERepr k f 'EUnit
+  EVarRepr :: f n -> MERepr k f ('EVar n)
+  EOpRepr ::
+    MERepr k f me1 ->
+    MERepr k f me2 ->
+    MERepr k f ('EOp me1 me2)
+
+type family Eval (k :: Type) (me :: MonoidExpr k) :: k where
+  Eval k 'EUnit = Unit k
+  Eval k ('EVar m) = m
+  Eval k ('EOp me1 me2) = Op k (Eval k me1) (Eval k me2)
+
+-- | Defunctionalization - inspired by \"singletons\", and with a similar naming
+-- scheme
+data TyFun a b
+type a ~> b = TyFun a b -> Type
+type family Apply (f :: a ~> b) (x :: a) :: b
+
+data IdSym1 :: MonoidExpr m ~> MonoidExpr m
+data EOpSym0 :: MonoidExpr m ~> (MonoidExpr m ~> MonoidExpr m)
+data EOpSym1 (me :: MonoidExpr m) :: MonoidExpr m ~> MonoidExpr m
```

Comment:
This ultimately comes down to personal taste, but I would prefer to write this as:

```suggestion
data EOpSym1 :: MonoidExpr m -> MonoidExpr m ~> MonoidExpr m
```

And similarly for other defunctionalization symbols that require at least one argument.

Reply:
Any particular reason/principle? Happy to change it either way, doesn't make a difference to me.

Reply:
Mainly because it makes the differences between the kinds of each defunctionalization symbol more apparent:

```hs
data EOpSym0 :: MonoidExpr m ~> MonoidExpr m ~> MonoidExpr m
data EOpSym1 :: MonoidExpr m -> MonoidExpr m ~> MonoidExpr m
```

Here, the difference between `EOpSym0` and `EOpSym1` is which arrow kind is used for the first argument.